README.md
# SleekCharts
[![Gem Version](https://badge.fury.io/rb/sleek_charts.png)](http://badge.fury.io/rb/sleek_charts) [![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/gouravtiwari/sleek_charts/trend.png)](https://bitdeli.com/free "Bitdeli Badge")

Provides Bar and Donut charts with consitent tooltip for Rails applications.
This gem uses [d3](http://d3js.org) and [d3-tip](https://github.com/Caged/d3-tip) libraries.

## Supports Rails 3.1+

![Bar tip](https://github.com/gouravtiwari/sleek_charts/raw/master/example/bar-tip.png)![Donut tip](https://github.com/gouravtiwari/sleek_charts/raw/master/example/donut-tip.png)

## Installation

Add this line to your application's Gemfile:

    gem 'sleek_charts'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install sleek_charts

Once done, add below to application's assets/javascripts/application.js file

    //= require sleek_charts

Also, add below to application's assets/stylesheets/application.css file

    *= require sleek_charts

## Usage

Simply add bar chart to any element, e.g.

    barTip(options);
    //where options is a map, checkout bar-tip.js
    //or with defaults:
    barTip({});
    

or add donut chart to any element, e.g.

    donutTip(options);
    //where options is a map, checkout donut-tip.js
    //or with defaults:
    donutTip({});

### Example

* [Audit Rails example](www.audit-rails.info/audit_rails/audits/analytics)

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request




CHANGELOG.md
### 0.0.3
- legends are optional for donut chart

### 0.0.1 & 0.0.2
- First and second minor releases with refactoring


app/assets/javascripts/bar-tip.js
// options: it is a map, which accepts below values
//
// selector:    selector to which bar chart will be added, default is 'body'
//              e.g. 'div#bar-chart'
// data:        data in JSON object, which contains label and value to create bar chart
//              e.g. [{'label': 'weather-day', 'value': 35.1}, {'label': 'weather-night', 'value': 30.2}]
// width:       width of svg element
// height:      height of svg element
// margin:      margin of svg element and accepts in a map, e.g.
//              {top: 40, right: 20, bottom: 120, left: 40}
// flexRight:   set it to false, when you are ok to display wider column, when there are <5-6 columns, default is true
// labelAngle:  default is 0
// xDomain:     x-axis domain, default is 'label'
// yDomain:     y-axis domain, default is 'value'
// yAxisText:   add additional text on y-axis default is ''
// tipLabel:    tipText which you want to display in tip, default is ''
// tipValue:    tipValue, for each bar, default is 'value'
// tipText:     tipText which you want to display with 'label' and 'value', default is ''

function mergeConfigOptions(defaults,options){
    var mergedConfig = {};
    for (var attrname in defaults) { mergedConfig[attrname] = defaults[attrname]; }
    for (var attrname in options)  { mergedConfig[attrname] = options[attrname]; }
    return mergedConfig;
}

function barTip(options){
    var defaults = {
        selector:   'body',
        data:       [{'label': 'weather-morning', 'value': 29.1}, {'label': 'weather-afternoon', 'value': 33.2},
                    {'label': 'weather-evening', 'value': 32.1}, {'label': 'weather-night', 'value': 30.2}],
        width:      600,
        height:     400,
        margin:     {top: 40, right: 20, bottom: 120, left: 40},
        flexRight:  true,
        labelAngle: -60,
        xDomain:    'label',
        yDomain:    'value',
        yAxisText:  'Temperature',
        tipLabel:   '',
        tipValue:   'value',
        tipText:    ''
    };

    var config = (options) ? mergeConfigOptions(defaults,options) : defaults;

    // Just to make sure the chart doesn't look clutter, when there are < 5-6 columns
    if(config.flexRight){
        config.margin['right'] = d3.max([20, (config.width - config.data.length * 50)]);
    }
        
    var width = config.width - config.margin.left - config.margin.right,
        height = config.height - config.margin.top - config.margin.bottom;

    var x = d3.scale.ordinal()
        .rangeRoundBands([0, width], .1);

    var y = d3.scale.linear()
        .range([height, 0]);

    var xAxis = d3.svg.axis()
        .scale(x)
        .orient("bottom");

    var yAxis = d3.svg.axis()
        .scale(y)
        .orient("left");

    var svg = d3.select(config.selector).append("svg")
        .attr("width", config.width)
        .attr("height", config.height)
      .append("g")
        .attr("transform", "translate(" + config.margin.left + "," + config.margin.top + ")");

    var tip = d3.tip()
      .attr('class', 'd3-tip')
      .offset([-10, 0])
      .html(function(d) {
        label = d[config.tipLabel] == undefined ? "" : " " + d[config.tipLabel];

        return "<strong>" + config.tipText + label + "</strong> <span style='color:red'>" + d[config.tipValue] + "</span>";
      });
    
    svg.call(tip);

    x.domain(config.data.map(function(d) { return d[config.xDomain]; }));
    y.domain([0, d3.max(config.data, function(d) { return d[config.yDomain]; })]);

    svg.append("g")
        .attr("class", "x axis")
        .attr("transform", "translate(0," + height + ")")
        .call(xAxis)
        .selectAll("text")
        .style("text-anchor", "end")
        .attr("dx", "-.8em")
        .attr("dy", ".15em")
        .attr("transform", function(d) {
            return "rotate("+ config.labelAngle +")" ;
        });

    svg.append("g")
        .attr("class", "y axis")
        .call(yAxis)
      .append("text")
        .attr("transform", "rotate(-90)")
        .attr("y", -40)
        .attr("x", -80)
        .attr("dy", "0.71em")
        .style("text-anchor", "end")
        .text(config.yAxisText);

    svg.selectAll(".bar")
        .data(config.data)
      .enter().append("rect")
        .attr("class", "bar")
        .attr("x", function(d) { return x(d[config.xDomain]); })
        .attr("width", x.rangeBand())
        .attr("y", function(d) { return y(d[config.yDomain]); })
        .attr("height", function(d) { return height - y(d[config.yDomain]); })
        .on('mouseover', tip.show)
        .on('mouseout', tip.hide);
}


app/assets/javascripts/donut-tip.js
// options: it is a map, which accepts below values
//
// selector:    selector to which donut chart will be added, default is 'body'
//              e.g. 'div#donut-chart'
// data:        data in JSON object, which contains label and value to create bar chart
//              e.g. [{'label': 'weather-day', 'value': 35.1}, {'label': 'weather-night', 'value': 30.2}]
// width:       width of svg element
// height:      height of svg element
// margin:      margin of svg element and accepts in a map
//              {top: 10, right: 20, bottom: 20, left:10}
// innerRadius: innerRadius for inner edge of donut
// outerRadius: outerRadius for outer edge of donut
// color:       color platter, default is d3.scale.category20c()
// xDomain:     x-axis domain, default is 'label'
// yDomain:     y-axis domain, default is 'value'
// tipLabel:    tipText which you want to display in tip, default is ''
// tipValue:    tipValue, for each bar, default is 'value'
// tipText:     tipText which you want to display with 'label' and 'value', default is ''
// totalLabel:  totalLabel to add a label for sum e.g. ' Views'
// legend:     default set to true to show legends

function mergeConfigOptions(defaults,options){
    var mergedConfig = {};
    for (var attrname in defaults) { mergedConfig[attrname] = defaults[attrname]; }
    for (var attrname in options)  { mergedConfig[attrname] = options[attrname]; }
    return mergedConfig;
}

function donutTip(options){
    var defaults = {
        selector:       'body',
        data:           [{'label': 'weather-morning', 'value': 29.1}, {'label': 'weather-afternoon', 'value': 33.2},
                        {'label': 'weather-evening', 'value': 32.1}, {'label': 'weather-night', 'value': 30.2}],
        width:          500,
        height:         500,
        margin:         {top: 10, right: 20, bottom: 20, left:10},
        innerRadius:    90,
        outerRadius:    180,
        color:          d3.scale.category20c(),
        xDomain:        'label',
        yDomain:        'value',
        tipLabel:       '',
        tipValue:       'value',
        tipText:        '',
        totalLabel:     '',
        legend:         true
    };

    var config = (options) ? mergeConfigOptions(defaults,options) : defaults;

    var width = config.width - config.margin.left - config.margin.right,
        height = config.height - config.margin.top - config.margin.bottom,
        total = d3.sum(config.data, function (d) {
            return d3.sum(d3.values(d));
        });

    var div = d3.select("body").append("div")
        .attr("class", "tooltip")
        .style("opacity", 0);

    var vis = d3.select(config.selector)
                .append("svg:svg") //create the SVG element inside the <body>    
                .data([config.data]) //associate our data with the document
                .attr("width", width + config.margin.left + config.margin.right)
                .attr("height", height + config.margin.top + config.margin.bottom)
                .append("svg:g") //make a group to hold our pie chart
                .attr("transform", "translate(" + config.outerRadius * 1.5 + "," + config.outerRadius * 1.5 + ")");
    
    var textTop = vis.append("text")
                .attr("dy", ".35em")
                .style("text-anchor", "middle")
                .attr("class", "textTop")
                .text("TOTAL")
                .attr("y", -10),
        textBottom = vis.append("text")
                .attr("dy", ".35em")
                .style("text-anchor", "middle")
                .attr("class", "textBottom")
                .text(total + " Views")
                .attr("y", 10);

    var arc = d3.svg.arc()
                .innerRadius(config.innerRadius)
                .outerRadius(config.outerRadius);
    var arcOver = d3.svg.arc()
                .innerRadius(config.innerRadius + 5)
                .outerRadius(config.outerRadius + 5);

    var pie = d3.layout.pie() //this will create arc data for us given a list of values
                .value(function (d) {
                    return d[config.yDomain];
                }); //we must tell it out to access the value of each element in our data array

    var arcs = vis.selectAll("g.slice") //this selects all <g> elements with class slice (there aren't any yet)
                .data(pie) //associate the generated pie data (an array of arcs, each having startAngle, endAngle and value properties) 
                .enter() //this will create <g> elements for every "extra" data element that should be associated with a selection. The result is creating a <g> for every object in the data array
                .append("svg:g") //create a group to hold each slice (we will have a <path> and a <text> element associated with each slice)
                .attr("class", "slice") //allow us to style things in the slices (like text)
                .on('mousemove', function (d, i) {
                    label = d[config.tipLabel] == undefined ? "" : " " + d[config.tipLabel];
                    d3.select(this).select("path").transition()
                    .duration(200)
                    .attr("d", arcOver);
                    
                    textTop.text(d3.select(this).datum().data[config.xDomain])
                    .attr("y", -10);
                textBottom.text(d3.select(this).datum().data[config.yDomain] + " " +config.totalLabel)
                    .attr("y", 10);
                    
                    div.transition()
                    .duration(200)
                    .style("opacity", 0.9);
                    div.html("<strong>" + config.tipText + label+ "</strong> <span style='color:red'>" + d.data[config.tipValue] + "</span>")
                    .style("left", (d3.event.pageX - 57) + "px")
                    .style("top", (d3.event.pageY - 50) + "px")
                    .style("z-index", 10000)
                })
                .on("mouseout", function (d) {
                    d3.select(this).select("path").transition()
                    .duration(100)
                    .attr("d", arc);
        
                    textTop.text("TOTAL")
                    .attr("y", -10);
                    textBottom.text(total + " " + config.totalLabel);
                    
                    div.transition()
                    .duration(500)
                    .style("opacity", 0);
                });

    arcs.append("svg:path")
        .attr("fill", function (d, i) {
            return config.color(i);
        }) //set the color for each slice to be chosen from the color function defined above
        .attr("d", arc); //this creates the actual SVG path using the associated data (pie) with the arc drawing function
    if(config.legend){
        var legend = d3.select(config.selector).append("svg")
            .attr("class", "legend")
            .attr("width", config.outerRadius)
            .attr("height", config.outerRadius * 2)
            .selectAll("g")
            .data(data)
            .enter().append("g")
            .attr("transform", function (d, i) {
            return "translate(0," + i * 20 + ")";
        });
        
        legend.append("rect")
            .attr("width", 18)
            .attr("height", 18)
            .style("fill", function (d, i) {
            return config.color(i);
        });
        
        legend.append("text")
            .attr("x", 24)
            .attr("y", 9)
            .attr("dy", ".35em")
            .text(function (d) {
                return d[config.xDomain];
        });
    }
}
